js2-mode.el

```diff
@@ -6074,7 +6078,9 @@ its relevant fields and puts it into `js2-ti-tokens'."
                           (?,
                            (throw 'return js2-COMMA))
                           (??
-                           (throw 'return js2-HOOK))
+                           (if (js2-match-char ?.)
+                               (throw 'return js2-OPTIONAL-CHAINING)
```

Comment:
Looking at the parsing function, we don't seem to save any information that the chaining was optional.

It seems like we'll get exactly the same result if we simply return `js2-DOT` here and skip all other changes.

Reply:
@dgutov , optional chaining is new JS syntax, see https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining

Reply:
I see. So maybe the patch makes sense.

But speaking of minimizing changes, you could look after `?.` and if the next char is `(`, return just that and ignore the preceding two characters. Again, if we don't store any info about optional chaining in the AST.

Reply:
I'm not sure I can hack the token stream. It's not the right way to do it.

Say user input `hello?.?.(` (that's wrong syntax, if I just remove `?.` after `(` from token stream. the remaining tokens becomes  `hello?.(`. So wrong usage of that operator now become syntax correct.

`use-optional-chaining-p` is not some hack. It strictly follows the standard workflow of parser,

Here is BNF of function call
FUNCTION: OPERATION  "(" [CALL_ARGS]")"
                  | OPERATION  OPTIONAL_CHAINING_OPERATOR  "(" [CALL_ARGS]")"

When `use-optional-chaining-p` is nil, pop OPERATION token only. It's the first scenario.
When `use-optional-chaining-p` is t, pop OPERATION token and one OPTIONAL_CHAINING_OPERATOR token. It corresponds to second scenario.

Reply:
OK, thanks for the explanation!